costmap_cspace/include/costmap_cspace/costmap_3d_layer/footprint.h

```diff
@@ -266,17 +294,48 @@ class Costmap3dLayerFootprint : public Costmap3dLayerBase
       }
     }
   }
+
+  const Rect& getDefaultRange(const nav_msgs::OccupancyGrid::ConstPtr&, const int) const
+  {
+    return template_ranges_[0];
+  }
+  const Rect& getMaskedRange(const nav_msgs::OccupancyGrid::ConstPtr& msg, const int pos) const
+  {
+    const int gx = pos % msg->info.width;
+    const int gy = pos / msg->info.width;
+    uint8_t mask = 0;
+    if (gx == 0 || (msg->data[pos - 1] >= msg->data[pos]))
+    {
+      mask += 1;
+    }
+    if (gx == static_cast<int>(msg->info.width) - 1 || (msg->data[pos + 1] >= msg->data[pos]))
+    {
+      mask += 2;
+    }
+    if (gy == 0 || (msg->data[pos - msg->info.width] >= msg->data[pos]))
+    {
+      mask += 4;
+    }
+    if (gy == static_cast<int>(msg->info.height) - 1 || (msg->data[pos + msg->info.width] >= msg->data[pos]))
+    {
+      mask += 8;
+    }
+    return template_ranges_[mask];
```

Comment:
As here are all conditions for template generation to calculate the template index, directly generating `Rect` here would reduce heap memory access without performance degradation